cmd/opentelemetry/app/receiver/kafkareceiver/kafka_receiver.go

```diff
@@ -17,79 +17,135 @@ package kafkareceiver
+		}
+	}
+
+	return cfg
+}
+
+// CreateTraceReceiver creates Jaeger receiver trace receiver.
+// This function implements OTEL component.ReceiverFactory interface.
+func (f *Factory) CreateTraceReceiver(
+	ctx context.Context,
 	params component.ReceiverCreateParams,
+	cfg configmodels.Receiver,
+	nextConsumer consumer.TraceConsumer,
 ) (component.TraceReceiver, error) {
-	w := &writer{receiver: config.Name(), nextConsumer: nextConsumer}
-	consumer, err := builder.CreateConsumer(
-		params.Logger,
-		metrics.NullFactory,
-		w,
-		config.Options)
-	if err != nil {
-		return nil, err
-	}
-	return &kafkaReceiver{
-		consumer: consumer,
-		logger:   params.Logger,
-	}, nil
+	return f.Wrapped.CreateTraceReceiver(ctx, params, cfg, nextConsumer)
 }
 
-// Start starts the receiver.
-func (r kafkaReceiver) Start(_ context.Context, _ component.Host) error {
-	r.consumer.Start()
-	return nil
+// CreateMetricsReceiver creates a metrics receiver based on provided config.
+// This function implements component.ReceiverFactory.
+func (f *Factory) CreateMetricsReceiver(
+	ctx context.Context,
+	params component.ReceiverCreateParams,
+	cfg configmodels.Receiver,
+	nextConsumer consumer.MetricsConsumer,
+) (component.MetricsReceiver, error) {
+	return f.Wrapped.CreateMetricsReceiver(ctx, params, cfg, nextConsumer)
 }
 
-// Shutdown shutdowns the receiver.
-func (r kafkaReceiver) Shutdown(_ context.Context) error {
-	return r.consumer.Close()
+// CreateLogsReceiver creates a receiver based on the config.
+// If the receiver type does not support logs or if the config is not valid
+// error will be returned instead.
+func (f Factory) CreateLogsReceiver(
+	ctx context.Context,
+	params component.ReceiverCreateParams,
+	cfg configmodels.Receiver,
+	nextConsumer consumer.LogsConsumer,
+) (component.LogsReceiver, error) {
+	return f.Wrapped.CreateLogsReceiver(ctx, params, cfg, nextConsumer)
 }
 
-// WriteSpan writes a span to the next consumer.
-func (w writer) WriteSpan(ctx context.Context, span *model.Span) error {
-	batch := model.Batch{
-		Spans:   []*model.Span{span},
-		Process: span.Process,
+// MustOtelEncodingForJaegerEncoding translates a jaeger encoding to a otel encoding
+func MustOtelEncodingForJaegerEncoding(jaegerEncoding string) string {
+	switch jaegerEncoding {
+	case kafka.EncodingProto:
+		return "jaeger_proto"
+	case kafka.EncodingJSON:
+		return "jaeger_json"
 	}
-	traces := jaegertranslator.ProtoBatchToInternalTraces(batch)
-	return w.nextConsumer.ConsumeTraces(w.addContextMetrics(ctx), traces)
-}
 
-// addContextMetrics decorates the context with labels used in metrics later.
-func (w writer) addContextMetrics(ctx context.Context) context.Context {
-	// TODO too many mallocs here, should be a cheaper way
-	ctx = obsreport.ReceiverContext(ctx, w.receiver, "kafka", "kafka")
-	ctx = obsreport.StartTraceDataReceiveOp(ctx, TypeStr, "kafka")
-	return ctx
+	panic(jaegerEncoding + " is not a supported kafka encoding in the OTEL collector.")
```

Comment:
I am not sure I like this but the config factory does not return an error

Reply:
Yes.  I did not like this either, but was unsure of a better way to handle it.  If this happens it would only be at startup which I think makes it acceptable.